File: 底层原理/底层原理-多线程/底层原理-多线程/ViewController.m
//
//  ViewController.m
//  底层原理-多线程
//
//  Created by weconex on 2021/7/1.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    NSLog(@"任务1");
    
    //主队列  先进先出
//    dispatch_queue_t queue = dispatch_get_main_queue();
//    //立马在当前线程执行任务 执行完毕继续往下执行
////    dispatch_sync(queue, ^{
////        NSLog(@"任务2");
////
////    });
//    //异步  在主队列中，等待上一个任务执行完开始执行
//    dispatch_async(queue, ^{
//        NSLog(@"任务2");
//
//    });
//    NSLog(@"任务3");
//
//    for ( int i = 0; i<1000; i++) {
//
//    }
    
//    NSLog(@"任务4");
    
    
    //串队列  先进先出
    dispatch_queue_t queue = dispatch_queue_create(@"myq", DISPATCH_QUEUE_CONCURRENT);
//    dispatch_queue_t queue2 = dispatch_queue_create(@"myq2", DISPATCH_QUEUE_SERIAL);
    //立马在当前线程执行任务 执行完毕继续往下执行
//    dispatch_sync(queue, ^{
//        NSLog(@"任务2");
//
//    });
    //异步  在主队列中，等待上一个任务执行完开始执行
    dispatch_async(queue, ^{
        NSLog(@"任务2");
        dispatch_sync(queue, ^{
            NSLog(@"任务3");
        });
        NSLog(@"任务4");
    });
    NSLog(@"任务5");
    
    for ( int i = 0; i<1000; i++) {
        
    }
    
    NSLog(@"任务6");
    
}
-(void)test{
    NSLog(@"2");
}

-(void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    //内部向runloop中添加定时器，子线程没有启动runloop  带有延迟的本质是向runloop中添加定时器
    [self performSelector:@selector(test) withObject:nil afterDelay:0.0];
}
@end

File: 底层原理/runloop-定时器失效的问题/runloop-定时器失效的问题/ViewController.swift
//
//  ViewController.swift
//  runloop-定时器失效的问题
//
//  Created by weconex on 2021/6/30.
//

import UIKit

class ViewController: UIViewController {

  final  var  stoped : Bool?
    var thead : ISThread?
    override func viewDidLoad() {
        super.viewDidLoad()
//        let mytimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { timer in
//            print("----")
//        }
//        let mytimer = Timer(timeInterval: 1.0, repeats: true) { timer in
//            print("-----")
//        }
////        RunLoop.main.add(mytimer, forMode: .tracking)
//        //common 不是一个真的模式，只是一个标记
//        // MARK:-- 在设置里common标记的模式下都能运行
//        RunLoop.main.add(mytimer, forMode: .common)
        let textView = UITextView(frame: CGRect(x: 10, y: 100, width: 200, height: 200))
        textView.text = "124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-1243254365124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-1243254365124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-124325436586797809-"
        view.addSubview(textView)
        
        
        /**
         线程保活
         */
        
    thead =  ISThread.init(target: self, selector: #selector(self.run), object: nil)
        thead?.start()
        
    }

    @objc  func run() {
        print("做事情")
        RunLoop.current.add(Port(), forMode: .default)
        RunLoop.current.run()
        
        print("end")
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.performSelector(onMainThread: #selector(self.test), with: nil, waitUntilDone: false)
        print("123")
    }
    
    @objc  func test() {
        print("test")
        
    }
}

